streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -465,6 +466,53 @@ public synchronized void onChange(final Thread thread,
         }
     }
 
+    final class DelegatingStateRestoreListener implements StateRestoreListener {
+        @Override
+        public void onRestoreStart(final TopicPartition topicPartition, final String storeName, final long startingOffset, final long endingOffset) {
+            if (globalStateRestoreListener != null) {
+                try {
+                    globalStateRestoreListener.onRestoreStart(topicPartition, storeName, startingOffset, endingOffset);
+                } catch (final Exception fatalUserException) {
+                    throw new StreamsException(
+                            String.format("Fatal user code error in store restore listener for store %s, partition %s.",
+                                    storeName,
+                                    topicPartition),
+                            fatalUserException);
+                }
+            }
+        }
+
+        @Override
+        public void onBatchRestored(final TopicPartition topicPartition, final String storeName, final long batchEndOffset, final long numRestored) {
+            if (globalStateRestoreListener != null) {
+                try {
+                    globalStateRestoreListener.onBatchRestored(topicPartition, storeName, batchEndOffset, numRestored);
+                } catch (final Exception fatalUserException) {
+                    throw new StreamsException(
```

Comment:
There is similar code forming StreamsException.
Consider refactoring to reduce code dup (can be done in another issue).